fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/RelationTransformer.java

```diff
@@ -406,16 +406,73 @@ private LogicalPlan processSetOperation(SetOperationRelation setOperationRelatio
+                                      SetOperationRelation setRelation) {
+        addProjection:
+        if (setRelation.hasOrderByClause()) {
+            for (OrderByElement orderByElement : setRelation.getOrderBy()) {
+                final Expr expr = orderByElement.getExpr();
+                // order by constant will be ignored
+                if (!(expr instanceof SlotRef) && !expr.isLiteral()) {
+                    break addProjection;
+                }
+            }
+            return root;
+        }
+
+        ExpressionMapping outputTranslations = new ExpressionMapping(root.getScope(), root.getFieldMappings(),
+                root.getColumnRefToConstOperators());
+        Map<ColumnRefOperator, ScalarOperator> projections = Maps.newHashMap();
+
+        for (ColumnRefOperator outputColumn : outputColumns) {
+            projections.put(outputColumn, outputColumn);
+        }
+
+        final List<Expr> orderByExpressions = setRelation.getOrderByExpressions();
+        for (Expr expression : orderByExpressions) {
+            final ScalarOperator scalarOperator = SqlToScalarOperatorTranslator.translate(expression,
+                    root.getExpressionMapping(), columnRefFactory);
+
+            ColumnRefOperator columnRefOperator = null;
+            if (scalarOperator.isColumnRef()) {
+                columnRefOperator = (ColumnRefOperator) scalarOperator;
+            } else if (scalarOperator.isVariable() && projections.containsValue(scalarOperator)) {
+                for (Map.Entry<ColumnRefOperator, ScalarOperator> entry : projections.entrySet()) {
+                    if (entry.getValue().equals(scalarOperator)) {
+                        columnRefOperator = entry.getKey();
+                        break;
+                    }
+                }
+            } else {
+                columnRefOperator =
+                        columnRefFactory.create(expression, expression.getType(), scalarOperator.isNullable());
+            }
+            projections.put(columnRefOperator, scalarOperator);
+            outputTranslations.put(expression, columnRefOperator);
+            if (scalarOperator.isConstant()) {
+                outputTranslations.putConstOperator(columnRefOperator, scalarOperator);
+            }
+        }
+
+        final LogicalProjectOperator projectOperator = new LogicalProjectOperator(projections);
+        root = new OptExprBuilder(projectOperator, Lists.newArrayList(root), outputTranslations);
+        return root;
+    }
+
     private OptExprBuilder addOrderByLimit(OptExprBuilder root, QueryRelation relation) {
         List<OrderByElement> orderBy = relation.getOrderBy();
         if (relation.hasOrderByClause()) {
             List<Ordering> orderings = new ArrayList<>();
             List<ColumnRefOperator> orderByColumns = Lists.newArrayList();
             for (OrderByElement item : orderBy) {
+                if (item.getExpr().isLiteral()) {
```

Comment:
order by "a"?

Reply:
What about `order by 1`, can it be replaced with the first column?

Reply:
> order by "a"?

order by "1"/ order by null will be ignored. the same process with QueryTransformer

Reply:
> What about `order by 1`, can it be replaced with the first column?

order by 1 here is the SlotRef.